### PR TITLE
Tag helper for image resizing

### PR DESCRIPTION
**Pull request by KeithRaven**
Added tag helper for razor to provide same functionality as the liquid image resize filters ([http://orchardcore.readthedocs.io/en/dev/OrchardCore.Modules/OrchardCore.Media/README/#image-resizing-filters](http://orchardcore.readthedocs.io/en/dev/OrchardCore.Modules/OrchardCore.Media/README/#image-resizing-filters))

Added example code to the docs.

### CODE REVIEW

**Comment by sebastienros** [comment]
Looks good, should we also adda `media-src` to do the asset url automatically?

**Comment by KeithRaven** [comment]
I wasn't sure.  It felt like the asset url resolution was maybe a separate concern although it would more user friendly to combine them.

Also I though there was probably a reason (that I don't know about :)) why this had been implemented via the AssetUrl extension method instead of a tag helper.

What do you think?

**Comment by sebastienros** [comment]
I initially thought you wanted to create razor helpers, not tag helpers. Now that I see it we could use the asset url in this tag helper. But the razor helper is still useful if we need it outside of an img tag.

**Comment by KeithRaven** [comment]
That makes sense. I will:
- Add media-src
- Extend the razor helper to allow for resize parameters
- Add a utility method for generating the url which they can both call into

Kind of related, you currently have the array of restricted sizes. This isn't currently working though as you can successfully request arbitrary sizes.  I debugged it and its failing on (OrchardCore.Media Startup.cs)
```
 if (Array.BinarySearch<int>(Sizes, parsedWidth) == -1)
                                        {
                                            validation.Commands.Clear();
                                        }
```
as Array.BinarySearch seems to return a variety of negative values if the value isn't found (I think it depends on how it compares to the items that _are_ in the array). Shall I do a seperate pull request for this?

With calls to generate the resize urls going through a central method, could we use this to register valid format requests (ie width, height, mode combinations) and then check against this in the onvalidate method of imagessharp?

I personally like image profiles but that will have to wait :)

**Comment by sebastienros** [comment]
Please use a separate PR in case that would block your previous work. And we could also have this from configuration, because that will have to be at some point.

**Comment by sebastienros** [comment]
Have you ensured the tag works if we only specify `media-src` without resizing?

File: src/OrchardCore.Modules/OrchardCore.Media/TagHelpers/ImageTagHelper.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+        public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            if (string.IsNullOrEmpty(Src))
+            {
+                return;
+            }
+
+            var resolvedSrc = _mediaFileStore != null ? _mediaFileStore.MapPathToPublicUrl(Src) : Src;
+            var resizedSrc = ImageSharpUrlFormatter.GetMediaResizeUrl(resolvedSrc, ImageWidth, ImageHeight, ResizeMode);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Media/TagHelpers/ImageTagHelper.cs]: If the `media-src` is not set then this shouldn't be called.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Media/TagHelpers/ImageTagHelper.cs]: I mean the `src` shouldn't be converted. Resizing and path conversion are independent.

**Comment by KeithRaven** [comment]
> Have you ensured the tag works if we only specify media-src without resizing?

I have now.  I previously missed adding the appropriate HtmlTargetElement attribute to allow this

File: src/OrchardCore.Modules/OrchardCore.Media/TagHelpers/ImageTagHelper.cs (hunk 1 above)
Comment (KeithRaven) [reply src/OrchardCore.Modules/OrchardCore.Media/TagHelpers/ImageTagHelper.cs]: I'm not sure if I've understood correctly but...

If media-src isn't set then the tag helper wont be called.  media-src can be set by itself with no resize attributes and so is independent of resizing.

I thought about making media-src a separate helper which fired first and so making the asset directory resolution and resizing completely separate, however realised that image resizing only occurs for images in the media directory and therefor felt that resizing should be dependant on the use of media-src.

Does that make sense/sound reasonable?

**Comment by sebastienros** [comment]
Coming back on this one, can you confirm that these will work:

Render an image from a local media path, and converts it to the public url
``

Use the `src` as is and resize the image
``

**Comment by KeithRaven** [comment]
>Render an image from a local media path, and converts it to the public url
``

Yes this currently works

>Use the src as is and resize the image
``

No it does not do this currently.  Sorry I previously added the explanation below under another of your comments but it got hidden due to code changes:

>If media-src isn't set then the tag helper wont be called. media-src can be set by itself with no resize attributes and so is independent of resizing (I've updated so that it explicitly returns before calling GetMediaResizeUrl if there are no dimensions set).

>I thought about making media-src a separate tag helper which fired first and so making the asset directory resolution and resizing completely separate, however realised that image resizing only occurs for images in the media directory and therefor felt that resizing _should_ be dependant on the use of media-src so that it could not be applied to urls not in the media directory.

If you would still like the src and resizing to be completely separate (i.e. work as per your second test case) then let me know and I will change it to do this.

**Comment by sebastienros** [comment]
I think it would be better. We could then use it on static files that are part of the theme or the modules.
If it's too complex then don't worry.

**Comment by KeithRaven** [comment]
Should be straightforward, I'll update it.

**Comment by KeithRaven** [comment]
I have updated so that asset url resolution is now independent of resizing.  It is possible to use:

` `

` `

` `

and any combination of width, height and mode.

Are you happy with 'media-' or should this be more consistent with other methods, e.g. 'asset-src'?  And if image resizing may include theme etc in the future should resizing become 'img-width' etc?

**Comment by sebastienros** [comment]
I think you two suggestions are good: `asset-src` and `img-width`. Then also make `asset-src` work outside of the img tag? But not the `witdh` and other image related attributes.

**Comment by sebastienros** [comment]
If anyone wants to implement the last comment to help @KeithRaven ? Otherwise I'll jump on it.

**Comment by KeithRaven** [comment]
Sorry been tied up on other things.  will push later on if that's ok.

By convention should these be prefixed with 'asp-' ? From a quick read it felt like they maybe should as they're on a native element.  Had you considered an orchard specific prefix to group the orchard tag helpers by for native elements?

**Comment by KeithRaven** [comment]
I checked the existing ones and the only one which applies to a native element was the ContentLinkTagHelper and this didn't use a prefix but thought I'd ask :)

**Comment by sebastienros** [comment]
No prefix, thanks.

**Pull request merged**